fc/mef.py

```diff
@@ -217,265 +52,152 @@ def clustering_gmm(data, n_clusters = 8, initialization = 'distance_sub',
+def selection_std(populations,
+                  low=None,
+                  high=None,
+                  n_std_low=2.5,
+                  n_std_high=2.5):
     """
-    Find the mode of a dataset by finding the peak of a smoothed histogram.
+    Select populations if most of their elements are between two values.
 
-    This function finds the mode of a dataset by calculating the histogram,
-    using a 1D Gaussian filter to smooth out the histogram, and identifying
-    the maximum value of the smoothed histogram. The ``sigma`` parameter of
-    the Gaussian filter is taken from the standard deviation of `data`.
+    This function selects populations from `populations` if their means are
+    more than `n_std_low` standard deviations greater than `low`, and
+    `n_std_high` standard deviations lower than `high`.
 
     Parameters
     ----------
-    data : array
-        Nx1 array to calculate the mode from. `data` is assumed to contain
-        only integers.
-    min_val : int, optional
-        Minimum possible value in `data`.
-    max_val : int, optional
-        Maximum possible value in `data`.
+    populations : list of arrays or FCSData objects
+        Populations to select or discard.
+    low, high : int or float, optional
+        Left and right threshold. If None, use 0.015 of the lowest value
+        and 0.0985 of the highest value in the first population's `domain`,
+        if available.
+    n_std_low, n_std_high : float, optional
+        Number of standard deviations from `low` and `high`, respectively,
+        that a population's mean has to be closer than to be discarded.
 
     Returns
     -------
-    peak : float
-        Mode of `data`.
-    hist_smooth : array
-        ``(max_val - min_val + 1)``-long array containing the smoothed
-        histogram.
+    selected_mask : boolean array
+        Set of flags indicating whether a population has been selected.
 
     """
-    # Calculate bin edges and centers
-    bin_edges = np.arange(min_val, max_val + 2) - 0.5
-    bin_edges[0] = -np.inf
-    bin_edges[-1] = np.inf
-    bin_centers = np.arange(min_val, max_val + 1)
-
-    # Identify peak
-    # Calculate sample mean and standard deviation
-    # mu = np.mean(data)
-    sigma = np.std(data)
-    # Calculate histogram
-    hist, __ = np.histogram(data, bin_edges)
-    # Do Gaussian blur on histogram
-    # We have found empirically that using one half of the distribution's 
-    # standard deviation results in a nice fit.
-    hist_smooth = scipy.ndimage.filters.gaussian_filter1d(hist, sigma/2.)
-    # Extract peak
-    i_max = np.argmax(hist_smooth)
-    peak = bin_centers[i_max]
-
-    return peak, hist_smooth
-
-def find_peaks_median(data):
-    """
-    Find the median of a data array.
+    # Default thresholds
+    if low is None:
+        low = 0.015*populations[0].domain(0)[0]
+    if high is None:
+        high = 0.985*populations[0].domain(0)[-1]
```

Comment:
Seems like this would break on a list of arrays where `domain` doesn't exist. Seems like you should query `populations[0]` for a `domain` attribute and raise an error if `low` and `high` are not specified *and* list is not of `FCSData` objects.

This also assumes it's appropriate to only use the first population's `domain`. More general solution would not make this assumption; calculate a `low` and `high` for each element in `populations`. In practice, they will probably always be the same, but again feels like an unnecessary assumption which can be pretty easily relaxed (might also be able to solve it recursively by calling `selection_std` on all elements in the array if it detects sequence-type).